actors/builtin/miner/miner_state.go

```diff
@@ -259,6 +273,138 @@ func (st *State) ComputeProvingSet(store adt.Store) ([]abi.SectorInfo, error) {
 	return sectorInfos, nil
 }
 
+func (st *State) addLockedFund(store adt.Store, currEpoch abi.ChainEpoch, pledgeAmount abi.TokenAmount) error {
+	Assert(pledgeAmount.GreaterThanEqual(big.Zero()))
+
+	lockedFundsArray := adt.AsArray(store, st.LockedFundsArray)
+
+	cliffStartEpoch := currEpoch + PledgeCliff
+	fullyVestedEpoch := currEpoch + PledgeCliff + PledgeVestingPeriod
+
+	for i := cliffStartEpoch; i < fullyVestedEpoch; i += VestIncrement {
+		var lockedFundEntry abi.TokenAmount
+		quantizedVestEpoch := getQuantizedEpoch(i)
+		found, err := lockedFundsArray.Get(epochKey(quantizedVestEpoch), &lockedFundEntry)
+
+		if err != nil {
+			return err
+		}
+
+		if !found {
+			lockedFundEntry = big.Zero()
+		}
+
+		vestAmount := computeVestAmount(pledgeAmount, cliffStartEpoch, fullyVestedEpoch, quantizedVestEpoch)
+		newLockedFundEntry := big.Add(lockedFundEntry, vestAmount)
+		lockedFundsArray.Set(epochKey(quantizedVestEpoch), &newLockedFundEntry)
+	}
+
+	st.LockedFundsArray = lockedFundsArray.Root()
+
+	return nil
+}
+
+var errorContinue = fmt.Errorf("success")
+
+func (st *State) slashLockedFund(store adt.Store, currEpoch abi.ChainEpoch, amountToSlash abi.TokenAmount) (abi.TokenAmount, error) {
+
+	lockedFundsArray := adt.AsArray(store, st.LockedFundsArray)
+	amountSlashed := big.Zero()
+	remainToSlash := amountToSlash
+
+	var lockedFund abi.TokenAmount
+	var toDelete []uint64
+	// Assume ordering in lockedFundsArray by CliffStartEpoch
+	err := lockedFundsArray.ForEach(&lockedFund, func(k int64) error {
+		if remainToSlash.GreaterThan(big.Zero()) {
+			if k >= int64(currEpoch) {
+				// slashable
+				slashingAmount := big.Max(remainToSlash, lockedFund)
+				lockedFund = big.Sub(lockedFund, slashingAmount)
+
+				amountSlashed = big.Add(amountSlashed, slashingAmount)
+				remainToSlash = big.Sub(remainToSlash, slashingAmount)
+				Assert(remainToSlash.GreaterThanEqual(big.Zero()))
+
+				if lockedFund.IsZero() {
+					// we can delete this entry
+					toDelete = append(toDelete, uint64(k))
+				}
+			}
+			return nil
+		} else {
+			return errorContinue
+		}
+	})
+
+	if err != nil {
+		if err != errorContinue {
+			return big.Zero(), err
+		}
+	}
+
+	// TODO: expose batch delete on AMT
```

Comment:
That's not a TODO in this repo. File an issue on AMT and change this to "If AMT exposed a batch delete, we could use that to save some writes here".